### PR TITLE
fix(cloudmap): fix CloudMap Service import, expose ECS CloudMap Service

### LINKED ISSUE #4286: AppMesh VirtualNode creation fails due to broken servicediscovery.Service.from_service_attributes call
**Tldr version**: It is not possible to create AppMesh VirtualNode in Fargate environment because CloudMap service is already created by `ecs.FargateService` and `servicediscovery.Service.from_service_attributes` call to look up that CloudMap service is broken

**Detailed description:** When trying to create AppMesh VirtualNode using `servicediscovery.Service.from_service_attributes` as `cloud_map_service` parameter for `appmesh.VirtualNode` call we get the following error:

`jsii.errors.JSIIError: Cannot read property 'namespaceName' of undefined`

Using method `servicediscovery.Service.from_service_attributes` to import existing CloudMap service rather than letting AppMesh to create a new service is common when working with Fargate. The reason is that the `ecs.FargateService` call is using parameter `cloud_map_options` to create a new CloudMap ServiceDiscovery Service. This is mandatory in order to register the FargateTasks with CloudMap. Since the CloudMap Service is created during creation of FargateService, the only option is to import existing CloudMap service when creating AppMesh VirtualNode. It is not possible to let AppMesh `VirtualNode` to create that CloudMap service as it has been already created by `ecs.FargateService`. Moreover the object type in `servicediscovery.Service` (required by `appmesh.VirtualNode`) is `IService` which is different to `ecs.FargateService` that is using `CloudMapOptions` type of object. So althought we need to reference the same CloudMap service, it has to be looked up instead of just referenced from previous step, however the lookup method is broken.

By checking source code in aws_cdk/aws_appmesh/__init__.py we can see that namespaceName is not included in `class VirtualNode` nor `class CfnVirtualNode` as parameter. Instead it is included in AwsCloudMapServiceDiscoveryProperty.
In the TypeScript version of the documentation the same call to create a new VirtualNode namespace_name is listed as required parameter:
https://docs.aws.amazon.com/cdk/api/latest/docs/aws-appmesh-readme.html
```
const node = mesh.addVirtualNode('virtual-node', {
  hostname: 'node-a',
  **namespace,**
  listeners: {
    portMappings: [
      {
        port: 8081,
        protocol: appmesh.Protocol.HTTP,
      },
    ],
    healthChecks: [
      {
        healthyThreshold: 3,
        intervalMillis: 5000, // minimum
        path: `/health-check-path`,
        port: 8080,
        protocol: appmesh.Protocol.HTTP,
        timeoutMillis: 2000, // minimum
        unhealthyThreshold: 2,
      },
    ],
  },
});
```

### Reproduction Steps
Step1: create Fargate service first along with CloudMap service (see cloud_map_options)
```
fargate_service = ecs.FargateService(self,
            <your service name>,
            service_name=<your service name>,
            task_definition=task_def,
            assign_public_ip=False,
            cloud_map_options=ecs.CloudMapOptions(
                 dns_record_type=servicediscovery.DnsRecordType.A,
                 dns_ttl=core.Duration.seconds(10),
                 name=short_service_name
            ),
            platform_version=ecs.FargatePlatformVersion.LATEST,
            security_group=<your sg>,
            cluster=tenant_cluster,
            desired_count=0,
            enable_ecs_managed_tags=True,
        )
```

Step 2: try to create AppMesh virtual node (behaviour is the same for both methods to create virtual node - that means by using add virtual node method or creating separated virtual node and referencing the appmesh)
```
        virtual_node = appmesh.VirtualNode(self,
            virtual_node_name,
            mesh=<your app mesh>,
            cloud_map_service=servicediscovery.Service.from_service_attributes(self,
                 <your service name>,
                 dns_record_type=servicediscovery.DnsRecordType.A,
                 routing_policy=servicediscovery.RoutingPolicy.MULTIVALUE,
                 service_arn=<name space arn>,
                 service_id=<name space id>,
                 service_name=<your service name>,
            ),
            listener=appmesh.VirtualNodeListener(
                health_check=appmesh.HealthCheck(
                    protocol=appmesh.Protocol.TCP,
                ),
                port_mapping=appmesh.PortMapping(
                    port=<your application ports>,
                    protocol=appmesh.Protocol.TCP
                ),
            ),
            virtual_node_name=virtual_node_name
        )
```

### Error Log

 ` File "/usr/local/lib/python3.7/site-packages/aws_cdk/aws_appmesh/__init__.py", line 3131, in add_virtual_node
    return jsii.invoke(self, "addVirtualNode", [id, props])
  File "/usr/local/lib/python3.7/site-packages/jsii/_kernel/__init__.py", line 104, in wrapped
    return _recursize_dereference(kernel, fn(kernel, *args, **kwargs))
  File "/usr/local/lib/python3.7/site-packages/jsii/_kernel/__init__.py", line 267, in invoke
    args=_make_reference_for_native(self, args),
  File "/usr/local/lib/python3.7/site-packages/jsii/_kernel/providers/process.py", line 346, in invoke
    return self._process.send(request, InvokeResponse)
  File "/usr/local/lib/python3.7/site-packages/jsii/_kernel/providers/process.py", line 316, in send
    raise JSIIError(resp.error) from JavaScriptError(resp.stack)
jsii.errors.JSIIError: Cannot read property 'namespaceName' of undefined`

### Environment

  - **CLI Version      : 1.9.0 **
  - **Framework Version:**
  - **OS               : Mac OS**
  - **Language         : Python**

### Other

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by rix0rrr**
- Fix the missing initialization of `namespace` on CloudMap's
  `Service.fromServiceAttributes()`.
- Expose the created CloudMap Service on ECS services so that
  `fromServiceAttributes()` doesn't need to be called at all.
- Fix a number of property initialization errors across the library.

Fixes #4286.

BREAKING CHANGE: `cloudmap.Service.fromServiceAttributes` takes a newly
required argument `namespace`.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts [hunk 1]
```diff
@@ -174,7 +181,7 @@ export class VirtualRouter extends VirtualRouterBase {
 /**
  * Interface with properties ncecessary to import a reusable VirtualRouter
  */
-interface VirtualRouterAttributes {
+export interface VirtualRouterAttributes {
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts]: If that becomes exported, but was not before... Should it be explicitly marked `@experimental`? Or is the whole module experimental?

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 2]
```diff
@@ -175,7 +175,7 @@ abstract class ProjectBase extends Resource implements IProject {
    * May be unset, in which case this Project is not configured to use a VPC.
    * @internal
    */
-  protected _connections: ec2.Connections;
+  protected _connections: ec2.Connections | undefined;
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts]: Why not `_connections?: ec2.Connections`?

Comment (rix0rrr) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts]: No particular reason. I like the explicitness of this.

File: packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts (hunk 1 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-appmesh/lib/virtual-router.ts]: The module is experimental. I also added `fromAttributes` which is new, needs this type to be public.

File: packages/@aws-cdk/aws-codebuild/lib/project.ts (hunk 2 above)
Comment (nija-at) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts]: I prefer the `?` operator, only because it's consistent with everywhere else.

**Pull request merged**